File: src/main/java/aurocosh/divinefavor/common/core/DivineFavorCreativeTab.java
package aurocosh.divinefavor.common.core;

import aurocosh.divinefavor.common.block.base.ModBlock;
import aurocosh.divinefavor.common.block.common.ModBlocks;
import aurocosh.divinefavor.common.constants.ConstMisc;
import aurocosh.divinefavor.common.item.base.ModItem;
import aurocosh.divinefavor.common.item.common.ModItems;
import net.minecraft.block.Block;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.Collection;

public class DivineFavorCreativeTab extends CreativeTabs {
    public static final DivineFavorCreativeTab INSTANCE = new DivineFavorCreativeTab();
    private NonNullList list;

    public DivineFavorCreativeTab() {
        super(ConstMisc.MOD_ID);
        //ConstMisc.MOD_ID

        //setNoTitle();
        //setBackgroundImageName(ConstResources.GUI_CREATIVE);
    }

    @Override
    public ItemStack createIcon() {
        return new ItemStack(ModBlocks.blockFastFurnace);
    }

//    @Override
//    public ItemStack getTabIconItem() {
//        return getIconItemStack();
//    }

    @Override
    public boolean hasSearchBar() {
        return false;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void displayAllRelevantItems(NonNullList<ItemStack> stacks) {
        list = stacks;

        Collection<ModBlock> blocks = ModBlocks.getBlocks();
        blocks.forEach(this::addBlock);

        Collection<ModItem> items = ModItems.getItems();
        items.forEach(this::addItem);
    }

    private void addItem(Item item) {
        item.getSubItems(this, list);
    }

    private void addBlock(Block block) {
        addItem(Item.getItemFromBlock(block));
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/core/creative_tabs/DivineFavorCreativeTab.kt
package aurocosh.divinefavor.common.core.creative_tabs

import aurocosh.divinefavor.common.core.ResourceNamer
import aurocosh.divinefavor.common.item.base.ModItem
import aurocosh.divinefavor.common.registry.ModRegistries
import net.minecraft.block.Block
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

open class DivineFavorCreativeTab(name: String, val itemGetter: () -> ModItem) : CreativeTabs(ResourceNamer.getFullName("tab", name).toString()) {
    private lateinit var list: NonNullList<ItemStack>

    override fun createIcon(): ItemStack {
//        return ItemStack(ModItems.grimoire)
        return ItemStack(itemGetter.invoke())
    }

    override fun hasSearchBar(): Boolean {
        return false
    }

    @SideOnly(Side.CLIENT)
    override fun displayAllRelevantItems(stacks: NonNullList<ItemStack>) {
        list = stacks

        ModRegistries.blocks.values.forEach(this::addBlock)
        ModRegistries.items.values.forEach(this::addItem)
        ModRegistries.arrows.values.forEach(this::addItem)
        stacks.sortWith(ModItemStackComparator())
    }

    private fun addItem(item: Item) {
        item.getSubItems(this, list)
    }

    private fun addBlock(block: Block) {
        addItem(Item.getItemFromBlock(block))
    }
}
//DivineFavor.MOD_ID
//setNoTitle();
//setBackgroundImageName(ConstResources.GUI_CREATIVE);